### ISSUE TITLE
Can't persist a nested Map

### ISSUE THREAD

**Issue by UnlimitedX** [typegoose/typegoose#678]
## Versions

- System: linux
- ts-node 10.7.0
- Node v14.18.1
- MongoDB 5.0.6 Community
- Typegoose 9.7.0
- Mongoose 6.2.4

## What is the Problem?

When I try to create a Document that contains a Nested Document which contain a Map, a Mongoose error is triggered :

`/home/ultdx/dev/services/test-typegoose/node_modules/mongoose/lib/schema/SubdocumentPath.js:246
    return value.validate(fn);
                 ^
TypeError: Cannot read property 'validate' of undefined
    at SubdocumentPath.doValidate (/home/ultdx/dev/services/test-typegoose/node_modules/mongoose/lib/schema/SubdocumentPath.js:246:18)
    at /home/ultdx/dev/services/test-typegoose/node_modules/mongoose/lib/document.js:2719:18
    at processTicksAndRejections (internal/process/task_queues.js:77:11)`

## Code Example

```
import { mongoose, getModelForClass, prop, index, modelOptions, Severity, PropType } from "@typegoose/typegoose";


@modelOptions({
    options: {
        allowMixed: Severity.ALLOW
    },
    schemaOptions: {
        collection: 'ResourceParameter',
        timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' }
    }
})
class ResourceParameter {

    @prop()
    public value: any; // Could be anything a String | Date | number | boolean even an Object or a List<any>

}


@modelOptions({
    options: {
        allowMixed: Severity.ALLOW
    },
    schemaOptions: {
        timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' },
    }
})
@index({ label: 1 }, { unique: true, partialFilterExpression: { label: { $type: "string", $exists: true, $gt: ["", null] } } })
abstract class Resource {

    @prop()
    public label: string;

    @prop({ type: () => ResourceParameter, _id: true }, PropType.MAP)
    public parameters: Map<string, ResourceParameter> = new Map<string, ResourceParameter>();

}

@index({ '$**': "text" })
@modelOptions({
    schemaOptions: {
        collection: 'ObjectNested',
    }
})
class ObjectNested extends Resource {
}

@index({ '$**': "text" })
@modelOptions({
    schemaOptions: {
        collection: 'ObjectMain',
    }
})
class ObjectMain extends Resource {

    @prop({ type: () => ObjectNested, _id: true })
    public object_nested: ObjectNested; // This is an object nested

    @prop({ type: () => [ObjectNested], _id: true })
    public object_nesteds: ObjectNested[]; // This is an object nested

}



export class Main {

    public async run() {

        await mongoose.connect("mongodb://localhost:27017/test-app");

        const ObjectMainModel = getModelForClass(ObjectMain);




        let objectMain = new ObjectMain();
        objectMain.label = "object main new";


        let resourceParameter1 = new ResourceParameter()
        resourceParameter1.value = "Hello I am a nested resource parameter";
        objectMain.parameters.set("param1", resourceParameter1); // TRIGGERS TypeError: Cannot read property 'validate' of undefined

        //Instantiating nested object
        objectMain.object_nested = new ObjectNested();
        objectMain.object_nested.label = "object nested 1";

        let resourceParameter2 = new ResourceParameter()
        resourceParameter2.value = "Hello I am a nested resource parameter";
        objectMain.object_nested.parameters.set("param2", resourceParameter2); // TRIGGERS TypeError: Cannot read property 'validate' of undefined when document objectMain is created below

        let document = await ObjectMainModel.create(objectMain);


        console.dir(JSON.parse(JSON.stringify(document)));
        /**
        return value.validate(fn);
                 ^
        TypeError: Cannot read property 'validate' of undefined
            at SubdocumentPath.doValidate (/home/ultdx/dev/services/test-typegoose/node_modules/mongoose/lib/schema/SubdocumentPath.js:246:18)
            at /home/ultdx/dev/services/test-typegoose/node_modules/mongoose/lib/document.js:2719:18
            at processTicksAndRejections (internal/process/task_queues.js:77:11)

        ts-node 10.7.0
        Node v14.18.1
        MongoDB 5.0.6 Community
        Typegoose 9.7.0
        Mongoose 6.2.4
         */


        process.exit(0);
    }

}

(new Main()).run();
```

## Do you know *why* it happens?

*no*

---

**Comment by hasezoey** [comment]
this does not seem like a typegoose problem to me, though it would need to be converted to mongoose code to make a issue in mongoose

Some Notes:

> `new ResourceParameter()`

it is recommended to not instantiate a typegoose class, use plain objects (will be converted later) or a new document from the model

note: mongoose supports doing maps like `[['key1', 'value1'], ['key2', 'value2']]` instead of having to define the map beforehand

> `class ObjectNested extends Resource {}`

it is recommended to add at least a dummy value, otherwise it may create problems sometimes (may equal to other "empty" classes)

**Comment by UnlimitedX** [comment]
Thanks for the prompt reply!

> Some Notes:
> new ResourceParameter()
> it is recommended to not instantiate a typegoose class, use plain objects (will be converted later) or a new document from the model

In this specific code example, there is no constructor nor methods. In our main project the class is not only used by Typegoose. There is a lot of inter-connectivity between classes.

Do you advise an instantiation like this instead? 
`let resourceParameter: ResourceParameter = new this.model(ResourceParameter);`

`this.model` is the current model of the class (here a ResourceParameter), when being in a Service (we are using the MVC pattern).


>     class ObjectNested extends Resource {}
> 
> it is recommended to add at least a dummy value, otherwise it may create problems sometimes (may equal to other "empty" classes)

Okay, thanks for your advice. 

I'm going to open a ticket on Mongoose with the associated mongoose code. :-) 
I'll be referencing the issue here.

**Comment by hasezoey** [comment]
> Do you advise an instantiation like this instead?

yes, a new instance of a model, resulting in a document or also possible is a POJO instead of any instance for mongoose creation
also i know not everything has a model (like subdocuments), so a POJO would be enough, instances of classes should also work, but is not recommended (because types may be wrong from the actual document)

**Comment by smolinari** [comment]
@UnlimitedX 

> Do you advise an instantiation like this instead?

Just to add to maybe help with your inner picture of what Typegoose is doing. An entity class does two things, it offers typing for TypeScript to work and it also is more like metadata to feed into Typegoose to be converted into Mongoose schema. At some point that schema will eventually become a Mongoose model for Mongoose to do its thing. If you use an instance created from an entity class on its own, it is basically useless in the context of working with Mongoose. 

> In our main project the class is not only used by Typegoose. There is a lot of inter-connectivity between classes.

An entity class is also not meant to be used to add or be a part of any kind of non-data-access business logic. It definitely should not be intertwined with any other non-entity classes. At best you can use inheritance between entity classes for code reuse. If you are using the entity classes for anything other than metadata for Typegoose, that is going a step too far. You should be using the Mongoose generated models inside your application's models for data access, leaving the entity classes, as noted, purely as metadata for Typegoose. 

To put it another way, if you created normal Mongoose schema objects to create your Mongoose models with, you wouldn't be using them for anything else in your application, right? 

I hope that makes sense and helps. 

Scott

**Comment by hasezoey** [comment]
i could reproduce this behavior in typegoose (https://github.com/typegoose/typegoose-testing/commit/55d0572bf9d406849ed3a37bbd359e6d366f187a) and in mongoose (https://github.com/typegoose/typegoose-testing/commit/27937df09cd6d3a78adc707daf487bdf3f838a24), i have also already created a mongoose issue: https://github.com/Automattic/mongoose/issues/11501

reproduction repository & branch: https://github.com/typegoose/typegoose-testing/tree/verify678

**Comment by hasezoey** [comment]
Update on the mongoose issue:
It is now fixed upstream (in mongoose), which is not released yet, but will be in 6.2.8 (which is normally within the next ~1 week)

this issue will be closed once 6.2.8 will be released, unless the issue persists / still is open for some questions (that i have missed)
